File: FSharpWithAkka/Account/AccountActor.fs
[<RequireQualifiedAccess>]
module AccountActor

open System
open Akkling

open Lib.Types
open BankTypes
open ActorUtil
open Bank.Transfer.Domain

let getInternalTransferActor =
   getChildActorRef<AccountMessage, BankEvent<TransferPending>>

let start
   (persistence: AccountPersistence)
   (broadcaster: AccountBroadcast)
   (mailbox: Actor<_>)
   (initialState: AccountState)
   =
   let actorName = (ActorMetadata.account initialState.EntityId).Name

   let rec handler (account: AccountState) (mailbox: Actor<AccountMessage>) =
      function
      | StartChildren id ->
         MaintenanceFeeActor.start
            persistence.loadAccountEvents
            //(fun _ -> DateTime.UtcNow.AddDays -30)
            //(fun _ -> TimeSpan.FromDays 30)
            (fun _ -> DateTime.UtcNow.AddMinutes -2)
            (fun _ -> TimeSpan.FromMinutes 2)
            mailbox
            id
            (Guid.NewGuid())

         ignored ()
      | Lookup _ ->
         mailbox.Sender() <! account
         ignored ()
      | StateChange cmd ->
         let validation = Account.stateTransition account cmd

         match validation with
         | Error(err) ->
            broadcaster.broadcastError err |> ignore
            printfn "%A: validation fail %A" actorName err
            become (handler account mailbox)
         | Ok((event, newState) as validationResult) ->
            try
               (event |> Envelope.unwrap |> persistence.save).Wait()
               broadcaster.broadcast validationResult |> ignore
            with err when true ->
               broadcaster.broadcastError err.Message |> ignore
               printfn "%A" err
               reraise ()

            match event with
            | TransferPending evt when
               evt.Data.Recipient.AccountEnvironment = RecipientAccountEnvironment.Internal
               ->
               let selection =
                  getInternalTransferActor
                     mailbox
                     (ActorMetadata.internalTransfer evt.EntityId).Name

               let aref =
                  match selection with
                  | None _ ->
                     InternalTransferRecipientActor.start
                        mailbox
                        persistence
                        evt.EntityId
                  | Some aref -> aref

               aref <! evt
            | TransferPending evt when
               evt.Data.Recipient.AccountEnvironment = RecipientAccountEnvironment.Domestic
               ->

               select mailbox ActorMetadata.domesticTransfer.Path
               <! (evt |> DomesticTransferRecipientActor.TransferPending)
            | _ -> ()

            become (handler newState mailbox)

   let aref =
      spawn mailbox actorName (initialState |> handler |> actorOf2 |> props)

   aref <! StartChildren initialState.EntityId
   aref

File: CSharpWithLanguageExt/Account/AccountActor.cs
using LanguageExt;
using static LanguageExt.Prelude;
using Echo;
using static Echo.Process;
using static System.Console;

using Lib;
using Lib.Types;
using Lib.BankTypes;
using Bank.Account.Domain;
using Bank.Transfer.Domain;
using Bank.Transfer.Actors;

namespace Bank.Account.Actors;

public record AccountRegistry(
   AccountPersistence persistence,
   AccountBroadcast broadcast
) {
   public Task<Option<AccountState>> Lookup(Guid id) {
      var process = "@" + AccountActor.PID(id);

      if (ActorUtil.IsAlive(process)) {
         var account = ask<AccountState>(process, new LookupCmd(id));
         return TaskSucc(Some(account));
      }

      return
         from acct in persistence.loadAccount(id)
         from pid in TaskSucc(Some(AccountActor.Start(persistence, broadcast, acct)))
         select acct;
   }
}

public static class AccountActor {
   public static ProcessId Start(
      AccountPersistence persistence,
      AccountBroadcast broadcaster,
      AccountState initialState
   ) {
      var pid = spawn<AccountState, Command>(
         PID(initialState.EntityId),
         () => initialState,
         (AccountState account, Command cmd) => {
            if (cmd is StartChildrenCmd) {
               MaintenanceFeeActor.Start(
                  persistence.loadAccountEvents,
                  //lookBackDate: () => DateTime.UtcNow.AddDays(-30),
                  //scheduledAt: () => TimeSpan.FromDays(30),
                  lookBackDate: () => DateTime.UtcNow.AddMinutes(-2),
                  scheduledAt: () => TimeSpan.FromMinutes(2),
                  cmd.EntityId
               );
               return account;
            }
            if (cmd is LookupCmd) {
               reply(account);
               return account;
            }

            var validation = account.StateTransition(cmd);

            return validation.Match(
               Fail: errs => {
                  broadcaster.broadcastError(errs.Head.Message);
                  WriteLine($"AccountActor: validation fail {errs.Head.Message}");
                  return account;
               },
               Succ: tup => {
                  var evt = tup.Event;

                  try {
                     persistence.save(evt).Wait();
                     broadcaster.broadcast(tup);
                  } catch (Exception err) {
                     broadcaster.broadcastError(err.Message);
                     WriteLine(err);
                     throw;
                  }


                  if (evt is TransferPending) {
                     var childName = TransferRecipientActor.ActorName;

                     if (!ActorUtil.IsAlive(child(childName))) {
                        TransferRecipientActor.Start(persistence);
                     }

                     tellChild(childName, (TransferPending) evt);
                  }

                  return tup.NewState;
               }
            );
         }
      );

      register(pid.Name, pid);
      tell(pid, new StartChildrenCmd(initialState.EntityId));
      return pid;
   }

   public static Unit SyncStateChange(Command cmd) {
      tell("@" + PID(cmd.EntityId), cmd);
      return unit;
   }

   public static Unit Delete(Guid id) {
      var pid = "@" + PID(id);
      kill(pid);
      WriteLine($"Killed process {pid}");
      return unit;
   }

   public static string PID(Guid id) => $"accounts_{id}";

   private record StartChildrenCmd(Guid id) : Command(id);
}

record LookupCmd(Guid Id) : Command(Id);
